File: src/main/java/net/loxal/soa/restkit/endpoint/ballot/PollResource.java
/*
 * Copyright 2014 Alexander Orlov <alexander.orlov@loxal.net>. All rights reserved.
 */

package net.loxal.soa.restkit.endpoint.ballot;

import net.loxal.soa.restkit.client.RESTClient;
import net.loxal.soa.restkit.endpoint.Endpoint;
import net.loxal.soa.restkit.model.ballot.Poll;

import javax.inject.Inject;
import javax.validation.Valid;
import javax.validation.constraints.NotNull;
import javax.ws.rs.DELETE;
import javax.ws.rs.GET;
import javax.ws.rs.POST;
import javax.ws.rs.PUT;
import javax.ws.rs.Path;
import javax.ws.rs.PathParam;
import javax.ws.rs.client.Entity;
import javax.ws.rs.container.AsyncResponse;
import javax.ws.rs.container.ContainerRequestContext;
import javax.ws.rs.container.Suspended;
import javax.ws.rs.core.Context;
import javax.ws.rs.core.MediaType;
import javax.ws.rs.core.Response;
import java.net.URI;
import java.util.concurrent.TimeUnit;

@Path(PollResource.RESOURCE_PATH)
public class PollResource extends Endpoint {
    private static final String RESOURCE_NAME = "poll";
    public static final String RESOURCE_PATH = "ballot/" + RESOURCE_NAME;

    @Inject
    private RESTClient<Poll> client;

    @POST
    public void create(@NotNull @Valid Poll poll, @Context ContainerRequestContext requestContext, @Suspended AsyncResponse asyncResponse) {
        asyncResponse.setTimeout(ASYNC_RESPONSE_TIMEOUT, TimeUnit.SECONDS);

        Response createdPoll = client.post(Entity.json(poll));
        final String id = extractIdOfLocation(createdPoll);

        URI entityLocation = URI.create(requestContext.getUriInfo().getRequestUri().toString() + URI_PATH_SEPARATOR + id);

        asyncResponse.resume(Response.fromResponse(createdPoll).location(entityLocation).build());
        LOG.info(requestContext.getMethod());
    }

    @Path(ID_PATH_PARAM_PLACEHOLDER)
    @DELETE
    public void delete(@NotNull @PathParam(ID_PATH_PARAM) String id, @Context ContainerRequestContext requestContext, @Suspended AsyncResponse asyncResponse) {
        asyncResponse.setTimeout(ASYNC_RESPONSE_TIMEOUT, TimeUnit.SECONDS);

        Response response = client.delete(Poll.class, id);

        asyncResponse.resume(Response.fromResponse(response).type(MediaType.APPLICATION_JSON_TYPE).build());
        LOG.info(requestContext.getMethod());
    }

    @Path(ID_PATH_PARAM_PLACEHOLDER)
    @GET
    public void retrieve(@NotNull @PathParam(ID_PATH_PARAM) String id, @Context ContainerRequestContext requestContext, @Suspended AsyncResponse asyncResponse) {
        asyncResponse.setTimeout(ASYNC_RESPONSE_TIMEOUT, TimeUnit.SECONDS);

        Response response = client.get(Poll.class, id);

        asyncResponse.resume(Response.fromResponse(response).build());
        LOG.info(requestContext.getMethod());
    }

    @Path(ID_PATH_PARAM_PLACEHOLDER)
    @PUT
    public void update(@NotNull @Valid Poll poll, @NotNull @PathParam(ID_PATH_PARAM) String id, @Context ContainerRequestContext requestContext, @Suspended AsyncResponse asyncResponse) {
        asyncResponse.setTimeout(ASYNC_RESPONSE_TIMEOUT, TimeUnit.SECONDS);

        Response updated = client.put(Entity.json(poll), id);

        asyncResponse.resume(Response.fromResponse(updated).build());
        LOG.info(requestContext.getMethod());
    }
}

File: src/main/kotlin/net/loxal/soa/restkit/endpoint/ballot/PollResource.kt
/*
 * Copyright 2015 Alexander Orlov <alexander.orlov@loxal.net>. All rights reserved.
 */

package net.loxal.soa.restkit.endpoint.ballot

import net.loxal.soa.restkit.client.RepositoryClient
import net.loxal.soa.restkit.endpoint.Endpoint
import net.loxal.soa.restkit.model.ballot.Poll
import java.util.concurrent.TimeUnit
import javax.validation.Valid
import javax.validation.constraints.NotNull
import javax.ws.rs.*
import javax.ws.rs.client.Entity
import javax.ws.rs.container.AsyncResponse
import javax.ws.rs.container.ContainerRequestContext
import javax.ws.rs.container.Suspended
import javax.ws.rs.core.Context
import javax.ws.rs.core.MediaType
import javax.ws.rs.core.Response

@Path(PollResource.RESOURCE_PATH)
class PollResource : Endpoint() {

    private var client: RepositoryClient<Poll> = RepositoryClient()

    @POST
    @Path(Endpoint.ID_PATH_PARAM_PLACEHOLDER)
    fun create(@NotNull @Valid poll: Poll,
               @PathParam(Endpoint.ID_PATH_PARAM) id: String,
               @Context requestContext: ContainerRequestContext,
               @Suspended asyncResponse: AsyncResponse) {
        asyncResponse.setTimeout(Endpoint.ASYNC_RESPONSE_TIMEOUT.toLong(), TimeUnit.SECONDS)

        val createdPoll = client.post(Entity.json<Poll>(poll), id = id)
        val entityLocation = requestContext.uriInfo.requestUri

        asyncResponse.resume(Response.fromResponse(createdPoll).location(entityLocation).build())
        Endpoint.LOG.info(requestContext.method)
    }

    @Path(Endpoint.ID_PATH_PARAM_PLACEHOLDER)
    @DELETE
    fun delete(@NotNull @PathParam(Endpoint.ID_PATH_PARAM) id: String, @Context requestContext: ContainerRequestContext, @Suspended asyncResponse: AsyncResponse) {
        asyncResponse.setTimeout(Endpoint.ASYNC_RESPONSE_TIMEOUT.toLong(), TimeUnit.SECONDS)

        val response = client.delete(Poll::class.java, id)

        asyncResponse.resume(Response
                .status(response.status)
                .type(MediaType.APPLICATION_JSON_TYPE)
                .entity(response.readEntity(String::class.java))
                .build())
        Endpoint.LOG.info(requestContext.method)
    }

    @Path(Endpoint.ID_PATH_PARAM_PLACEHOLDER)
    @GET
    fun retrieve(@NotNull @PathParam(Endpoint.ID_PATH_PARAM) id: String, @Context requestContext: ContainerRequestContext, @Suspended asyncResponse: AsyncResponse) {
        asyncResponse.setTimeout(Endpoint.ASYNC_RESPONSE_TIMEOUT.toLong(), TimeUnit.SECONDS)

        val response = client.get(Poll::class.java, id)

        asyncResponse.resume(Response
                .status(response.status)
                .type(response.mediaType)
                .entity(response.readEntity(String::class.java))
                .build())
        Endpoint.LOG.info(requestContext.method)
    }

    @Path(Endpoint.ID_PATH_PARAM_PLACEHOLDER)
    @PUT
    fun update(@NotNull @Valid poll: Poll, @NotNull @PathParam(Endpoint.ID_PATH_PARAM) id: String, @Context requestContext: ContainerRequestContext, @Suspended asyncResponse: AsyncResponse) {
        asyncResponse.setTimeout(Endpoint.ASYNC_RESPONSE_TIMEOUT.toLong(), TimeUnit.SECONDS)

        val updated = client.put(Entity.json<Poll>(poll), id)

        asyncResponse.resume(Response.fromResponse(updated).build())
        Endpoint.LOG.info(requestContext.method)
    }

    companion object {
        private val RESOURCE_NAME = "poll"
        public val RESOURCE_PATH: String = "ballot/" + RESOURCE_NAME
    }
}